### ISSUE TITLE
Workflow metadata container attribute is not resolved when it's defined dynamically

### ISSUE THREAD

**Issue by MaxUlysse** [nextflow-io/nextflow#602]
cf https://github.com/SciLifeLab/NGI-RNAseq/pull/193
Using `container = wf_container` does work well for our docker containers.
But when we wanted to use singularity, we tried using `container = {"docker://$wf_container"}`, and we have a single `docker://` string instead of our container.
Do you think it'll be possible to use a closure regarding the type of container used (docker or singularity), or is that a little too far stretched?

**Comment by pditommaso** [comment]
could you provide the links where you are defining `wf_container` and where are you using it?

**Comment by MaxUlysse** [comment]
Sorry about that.
`wf_container` is defined in `nextflow.config`: https://github.com/SciLifeLab/NGI-RNAseq/blob/master/nextflow.config#L15
And used in `docker.config` : https://github.com/SciLifeLab/NGI-RNAseq/blob/master/conf/docker.config#L19
But the only way I manage to make it work, was by redefining everything back in our `singularity.config`: https://github.com/SciLifeLab/NGI-RNAseq/blob/master/conf/singularity.config#L19 which does not refer to this `wf_container` anymore.

**Comment by pditommaso** [comment]
So, the problem you are reporting is with the singularity configuration, right? could you please try to run it with this command line 

```
nextflow -trace nextflow.container run  SciLifeLab/NGI-RNAseq  -profile singularity
```

and attach the produced `.nextflow.log` file?

**Comment by ewels** [comment]
The cluster is down for maintenance but we'll try when it's up again. The Travis tests are interesting though... It looks like the container path is not defined [here](https://travis-ci.org/SciLifeLab/NGI-RNAseq/jobs/337999609#L2902) (`workflow.container` from [this line](https://github.com/SciLifeLab/NGI-RNAseq/blob/master/main.nf#L262)) but then it is just after [here](https://travis-ci.org/SciLifeLab/NGI-RNAseq/jobs/337999609#L2912)..

I can't see how that could happen?

**Comment by pditommaso** [comment]
It doesn't need to run. I need that log to see what's happening behind the scene. I've tried to run on my computer but I don't have the proper setup so it stop immediately.

**Comment by ewels** [comment]
Ok, @Hammarn should be able to run it locally and send it over. To run the tests we use [this script](https://github.com/SciLifeLab/NGI-RNAseq/blob/master/tests/container_test_hisat.sh) which should do everything:

```bash
cd tests
./container_test_hisat.sh --profile singularity
```

**Comment by Hammarn** [comment]
Ok, now that our clusters are starting to come back up again I'm able to actually test this.
I'm attaching the appropriate logfile
[nextflow.log](https://github.com/nextflow-io/nextflow/files/1710328/nextflow.log)
:

**Comment by pditommaso** [comment]
The image name looks fine.

```
  Failed to pull singularity image
  command: singularity pull --name scilifelab-ngi-rnaseq-latest.img docker://scilifelab/ngi-rnaseq:latest > /dev/null
  message:
  WARNING: pull for Docker Hub is not guaranteed to produce the
  WARNING: same image on repeated pull. Use Singularity Registry
  WARNING: (shub://) to pull exactly equivalent images.
  WARNING: Building container as an unprivileged user. If you run this container as root
  WARNING: it may be missing some functionality.
```

What version of singularity are you using? I fear they have messed up the pull command from docker (one more time).

**Comment by Hammarn** [comment]
```
singularity --version
2.4.2-dist
```

**Comment by pditommaso** [comment]
Try to run that command and see if it provides more info.

**Comment by Hammarn** [comment]
Now this was interesting:
```
singularity pull --name scilifelab-ngi-rnaseq-latest.img docker://scilifelab/ngi-rnaseq:latest
WARNING: pull for Docker Hub is not guaranteed to produce the
WARNING: same image on repeated pull. Use Singularity Registry
WARNING: (shub://) to pull exactly equivalent images.
Docker image path: index.docker.io/scilifelab/ngi-rnaseq:latest
Cache folder set to /pica/h1/richam/.singularity/docker
Importing: base Singularity environment
Importing: /pica/h1/richam/.singularity/docker/sha256:723254a2c089166d4bcfa917be0181ddbecd94971ebfe85792d96e7e29be9c68.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:abe15a44e12f84b4aac20d4f2b450ec8638cc0b176c9130d1802cb4fed17d953.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:409a28e3cc3de08700c9d37c809a6d3aa43dc076402943919f4a78c286c60a0b.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:50316693559041b3c6895bce352b10c302f31c074495671a820b689d2ce12baa.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:043a12c29ea469f2521637d790e447a11583651eaa1809da1b15cbbdcdccc0ed.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:3036204524472a3bf48d87444c797e8d0ea5e2b7b7490702a059314d797aa697.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:c61f95baa024b8c0a05ab25a2180f5a37d9cae8284ff70709ca0fcb36d97d9e6.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:3f2018472a1fdc5a8ef25fbbf2cde838d8c9aa99ca9e799c0f38d82552368d8c.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:a25f8a69c882f7cd39cfd819ae28c9d228bfc4255fe39a18a5d6bd41fd60b061.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:b80065da7bf43f0d5bb44441f2a8ea4f7a0dc9059118f40e10b5dc3cad21b620.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:6c41c3cb1e143813c55919a792447e3b73568bdf824cda9a30be0b72193329d2.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:370e0e0331b95a2ed8a5612ecf8733267a3d6c416e3eb80af67d0b2240d57d8c.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:7dd17cc9b029bd6480940a240eb38eb42f146be729632d5351648b01e13bd16c.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:aaae21d06ea0ec73fd47ded7bc388fbbe982e8340002becba469676650790202.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:322f88e64530eafcbb242d2286ca3799c4977965245fa9ed1158622f1a24eaeb.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:32fb7a5d2e2601b7c7410e643e3243ed01351d99f59371873942f0c3303879a5.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:2e27c7796aff580cd1b9a51449efeee0c3f0cf82df1c3683fa2689eb05cd83a9.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:022a479b955eca9183614f7ba2cc80123ae85271cd6016133224f80a5d1f5e73.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:d627b73660ad77aacb6dab2c49c275f7c974eca4cb9dde70b326239ce3646ad3.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:19fd7f95665d84598b61bdf5750d0f68b11de8c7774c4c75e73573a695d4238e.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:10bcc2cd5bca8722ee16d3bd30e79e6f46765aa9a5fb2adc610715524332b45a.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:57cb97e980641661cfed2a3ac8bb8c1fc49105fc81dc8c84dec118b7b1082841.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:36c38486bcf459ba981360f2da611310d98561da2791ca394bb6015eeecfb1a8.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:f4064e5427d824171633f05d9521ce53ec70d53efc597a9462828c9c5ac398ad.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:8db43d25192793d7a0280bd0329fdb6b0fbeb19ded8f4c42f798f3029e5d855b.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:655b2e723061001971c1af5102b5bbb9fd5586b47dcb96084ea4f1046300edcd.tar.gz
Importing: /pica/h1/richam/.singularity/docker/sha256:698daeb4d9e38b45bbbf210b2d8ab1ae9f86d32c6361af5fd9f4b135d6a8e619.tar.gz
Importing: /pica/h1/richam/.singularity/metadata/sha256:cc93d20166cb1040a76ca2d8b36374899844e9e9d206875a5d5e96a51b606f29.tar.gz
WARNING: Building container as an unprivileged user. If you run this container as root
WARNING: it may be missing some functionality.
Building Singularity image...
Singularity container built: ./scilifelab-ngi-rnaseq-latest.img
Cleaning up...
Done. Container is at: ./scilifelab-ngi-rnaseq-latest.img
```

**Comment by pditommaso** [comment]
Can you check the `singularity pull` exit status just after the execution (``echo $?)? it looks it's returning a non-zero.

**Comment by Hammarn** [comment]
No it seems to be exiting fine
```
Done. Container is at: ./scilifelab-ngi-rnaseq-latest.img
 » echo $?
0
```

**Comment by pditommaso** [comment]
I need to run a test with the sing version you are using. In the meanwhile you can pull the container manually and specify the image file as workaround.

**Comment by pditommaso** [comment]
What NF version are you using?

**Comment by pditommaso** [comment]
👍

**Comment by ewels** [comment]
I don't think that this is a singularity problem - it's more that the NF pipeline is not specifying an image name..? See the travis logs above, it's a blank value.

**Comment by pditommaso** [comment]
are you using singularity with travis ?

**Comment by ewels** [comment]
Yup, there's a singularity test and a docker test

**Comment by pditommaso** [comment]
In the travis log it seems the `workflow.container` is not reported but the right singularity is used.

**Comment by pditommaso** [comment]
I've tried to run the pipeline with `2.4.2-dist` and it works fine. The only problem that I see is that the `workflow.container` is not populated correctly when the the container is defined dynamically.

**Comment by ewels** [comment]
Yup, which ties in with what happens in the Travis test 👍 Rickard - the stuff you ran on Uppmax was not getting the container name for the processes though right? We were getting errors because of filenames like `.sing` and other weird stuff. Do you have the logs from those runs?

**Comment by pditommaso** [comment]
To recap, the container is not correctly displayed in the `workflow` metadata when it's  defined dynamically as a closure. The commit I've just pushed solve the issue. Since the travis build is running fine, the other problem seem more a issue on your side, therefore I'm closing this issue as resolved.

**Comment by ewels** [comment]
Ok brilliant - thanks @pditommaso! We'll ping you again if we can find something reproducible 👍

**Comment by pditommaso** [comment]
Fixed in version `0.27.5`.
